### PR TITLE
feat: reconcile 2.0, event streams, version enforcement

### PR DESCRIPTION
**Pull request by Meshiest**
tasks:
- [x] dataformat for agent status
- [x] dataformat for env info
- [x] storage of loki url
- [x] storage of resolved addrs
- [x] refactor resolved addrs to support agent IP updates
- [x] refactor agent reconcile to support non-rpc reconciliation
- [x] upgrade agent status to reflect current reconcile status
- [x] refactor resolved binaries to support binary updates
- [x] add new event stream for properly monitoring new agent reconcile status output
- [x] refactor cannon events to use new events
- [x] websocket api for event stream
- [x] CLI uses event stream to monitor reconcile/transaction statuses
- [x] fix transactions not using the right cost calculation
- ~~[ ] potentially start tests on reconcile 2.0 behavior~~ (need to make a new task for making all of snops more testable, with the controlplane not needing to run axum for each api to be tested)
- [x] rename prepare to apply

related:
- completes https://github.com/monadicus/snarkops/issues/220
- completes https://github.com/monadicus/snarkops/issues/181
- completes https://github.com/monadicus/snarkops/issues/141
- completes https://github.com/monadicus/snarkops/issues/126
- reimplements https://github.com/monadicus/snarkops/issues/206
- addresses https://github.com/monadicus/snarkops/issues/204 with new reconcile logic
- completes https://github.com/monadicus/snarkops/issues/187
- closes https://github.com/monadicus/snarkops/issues/188
- completes https://github.com/monadicus/snarkops/issues/279
- completes https://github.com/monadicus/snarkops/issues/77 by both event monitoring and improved reconcile monitoring
- enables https://github.com/monadicus/snarkops/issues/143 by event monitoring
- completes https://github.com/monadicus/snarkops/issues/192

### CODE REVIEW

File: crates/agent/src/main.rs [hunk 1]
```diff
@@ -149,232 +72,101 @@ async fn main() {
...
+                })
+                .unwrap_or_default(),
+        ),
+        resolved_addrs: RwLock::new(
+            db.resolved_addrs()
+                .inspect_err(|e| {
+                    error!("failed to load resolved addrs from db: {e}");
+                })
+                .unwrap_or_default(),
+        ),
```
[the remark this answers is not in the document]
Comment (Meshiest) [reply crates/agent/src/main.rs]: resolving from discussion - sled is mostly just for persistence rather than state keeping

File: crates/agent/src/reconcile/process.rs [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+                // update the kill time if the kill was successful
+                self.sigkill_at = Some(Instant::now());
+            })
+            .is_ok()
+    }
+}
+
+/// The EndProcessReconciler will return true when the child process has exited.
+/// It will wait NODE_GRACEFUL_SHUTDOWN_TIMEOUT seconds after sending a SIGINT
+/// before sending a SIGKILL (if the childi process has not exited),
```
Comment (gluax) [inline crates/agent/src/reconcile/process.rs:135]: childi

File: crates/agent/src/reconcile/storage.rs [hunk 3]
```diff
@@ -0,0 +1,517 @@
...
+
+/// Download a specific binary file needed to run the node
+pub struct BinaryReconciler<'a> {
+    pub state: Arc<GlobalState>,
+    pub env_info: Arc<AgentEnvInfo>,
+    pub node_binary: Option<InternedId>,
+    /// Metadata about an active binary transfer
+    pub transfer: &'a mut Option<(TransferId, BinaryEntry)>,
+    /// Time the binary was marked as OK
+    pub ok_at: &'a mut Option<Instant>,
```
Comment (gluax) [inline crates/agent/src/reconcile/storage.rs:33]: The preference for lifetimes on Options is to never have a reference to an Option, but rather the insides of an Option.

File: crates/agent/src/reconcile/storage.rs [hunk 4]
```diff
@@ -0,0 +1,517 @@
...
+}
+
+/// Download the genesis block needed to run the node
+pub struct GenesisReconciler<'a> {
+    pub state: Arc<GlobalState>,
+    pub env_info: Arc<AgentEnvInfo>,
+    /// Metadata about an active genesis transfer
+    pub transfer: &'a mut Option<TransferId>,
+    /// Time the genesis was marked as OK
+    pub ok_at: &'a mut Option<Instant>,
```
Comment (gluax) [inline crates/agent/src/reconcile/storage.rs:123]: Same here

File: crates/agent/src/reconcile/storage.rs [hunk 5]
```diff
@@ -0,0 +1,517 @@
...
+
+pub type LedgerModifyResult = Result<bool, ReconcileError>;
+
+pub struct LedgerReconciler<'a> {
+    pub state: Arc<GlobalState>,
+    pub env_info: Arc<AgentEnvInfo>,
+    pub target_height: (usize, HeightRequest),
+    pub last_height: &'a mut Option<(usize, HeightRequest)>,
+    pub pending_height: &'a mut Option<(usize, HeightRequest)>,
+    pub modify_handle: &'a mut Option<(AbortHandle, Arc<Mutex<Option<LedgerModifyResult>>>)>,
```
Comment (gluax) [inline crates/agent/src/reconcile/storage.rs:205]: Here.

File: crates/agent/src/state.rs [hunk 6]
```diff
@@ -76,49 +90,148 @@ impl GlobalState {
...
         }
     }
 
-    pub fn transfer_tx(&self) -> TransferTx {
-        self.transfer_tx.clone()
+    pub async fn set_node_status(&self, status: Option<SnarkOSStatus>) {
+        *self.last_node_status.write().await = status.map(|s| (Instant::now(), s));
+    }
+
+    pub async fn get_node_status(&self) -> Option<SnarkOSStatus> {
```
Comment (gluax) [inline crates/agent/src/state.rs:234]: Couldn't this be a reference? Or is needed as a non-reference where it's called?

File: crates/common/src/rpc/codec.rs [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+// the overhead of JSON for messages is not a concern for the RPC layer.
+
+pub fn encode<T: serde::Serialize>(msg: &T) -> serde_json::Result<Vec<u8>> {
+    serde_json::to_vec(msg)
+}
+
+pub fn decode<'de, T: serde::Deserialize<'de>>(msg: &'de [u8]) -> serde_json::Result<T> {
+    serde_json::from_slice(msg)
+}
+
```
Comment (gluax) [inline crates/common/src/rpc/codec.rs:11]: Do we still need these if we are removing `bincode`? Are we also full swapping from `tarpc` to a json one? Or will that be another PR?

**Review by gluax — APPROVED** [review]
Tried my best to go through it all catching stuff but big.

File: crates/agent/src/reconcile/storage.rs (hunk 3 above)
Comment (Meshiest) [reply crates/agent/src/reconcile/storage.rs:33]: These are mutable references to options because it's mutating data from the parent state

File: crates/agent/src/state.rs (hunk 6 above)
Comment (Meshiest) [reply crates/agent/src/state.rs:234]: I think it would have to be the RwLock's guard type, which felt kinda weird. Any references would be temporary

File: crates/agent/src/reconcile/storage.rs (hunk 3 above)
Comment (gluax) [reply crates/agent/src/reconcile/storage.rs:33]: Are you replacing the option from Some to None? Otherwise the inside being mutable would the way to go.

Comment (Meshiest) [reply crates/agent/src/reconcile/storage.rs:33]: yes

File: crates/common/src/rpc/codec.rs (hunk 7 above)
Comment (Meshiest) [reply crates/common/src/rpc/codec.rs:11]: We can swap to a real jsonrpc in another PR

File: crates/agent/src/reconcile/process.rs (hunk 2 above)
Comment (Meshiest) [reply crates/agent/src/reconcile/process.rs:135]: Gonna pretend this one is deliberate to thwart plagiarism without proper licensing rather than making a commit for one letter
